### PR TITLE
[Merged by Bors] - feat(CategoryTheory): the adjunction between Over.map and Over.baseChange

### PR DESCRIPTION
**Pull request by emilyriehl**
When a category has pullbacks, any morphism gives rise to an adjunction between over categories, with the left adjoint defined by composition and the right adjoint defined by pullback. This PR contains a definition of that adjunction.

---

This commit includes a single new definition completing a TODO listed at the end of the file. This is my first mathlib PR and I'm a new Lean user so I am sure there are many things that can be done better. Comments very welcome. 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by joelriou** [comment]
The definitions looks good, but here are some remarks which may improve the code:

-  `Over` categories have a convenient constructor for morphisms `Over.homMk` which allows to provide only the relevant morphism and the commutativity of a triangle, which eases the definitions.
- When defining data, in most cases, we should refrain from using the tactic mode (like `simp` in your definitions of the `left` fields), because it would usually make the given values of the data more difficult to use later.
- `eqToHom` should be used only when it is absolutely necessary (here, the identity map `𝟙 _` would have worked).
- the fields `homEquiv_naturality_...` can be completely omitted: in the category theory API, by default, Lean tries `by aesop_cat` in order to fill the missing proofs.
- the name `hasPullbackOverAdj` may not be the best, I would suggest changing it to `Over.mapAdjunction`, in the `Over` namespace. Similarly, I believe that the previously defined definition `baseChange f : Over Y ⥤ Over X` should be moved to the `Over` namespace (instead of the `Limits` namespace).

Taking these remarks into account, the code may roughly look like this:

```lean
/-- The adjunction `Over.map f ⊣ baseChange f`. -/
def Over.mapAdjunction [HasPullbacks C] {X Y : C} (f : X ⟶ Y) :
    Over.map f ⊣ baseChange f :=
  Adjunction.mkOfHomEquiv
    { homEquiv := fun x y =>
        { toFun := fun u => Over.homMk (pullback.lift u.left x.hom (by simp))
          invFun := fun v => Over.homMk (v.left ≫ pullback.fst) (by
            dsimp
            rw [Category.assoc, pullback.condition, ← Over.w v, Category.assoc]
            dsimp)
          left_inv := by aesop_cat
          right_inv := fun f => by
            dsimp
            ext
            dsimp
            ext
            . simp
            · dsimp
              rw [pullback.lift_snd, ← Over.w f]
              dsimp } }
```

**Comment by emilyriehl** [comment]
Thanks @joelriou. Is the correct workflow for me to now try to implement all of your suggestions?

**Comment by joelriou** [comment]
> Thanks @joelriou. Is the correct workflow for me to now try to implement all of your suggestions?

Yes. And then, you may change the label `awaiting-author` by `awaiting-review`. If necessary, do not hesitate asking questions here or on Zulip?

**Comment by emilyriehl** [comment]
Thanks. In that case, I have three questions:

(i) I noticed you moved the "{" from, e.g., the end of the line "Adjunction.mkOfHomEquiv" to the start of the following line (with suitable indentation). Is this preferred, even though it further indents subsequent code?

(ii) Is there an automatic way to convert a tactic style proof to a term style proof (with tactics listed in the "(by ...)" after the fact (perhaps in simple cases)? Eg, it's hard for me to image discovering your definition of "invFun" without being in tactic mode.

(iii) I only have a vague understanding of name spaces. In addition to the renaming you suggest, are you suggesting I move this definition and the previous one to the "Over" file?

**Comment by joelriou** [comment]
> (i) I noticed you moved the "{" from, e.g., the end of the line "Adjunction.mkOfHomEquiv" to the start of the following line (with suitable indentation). Is this preferred, even though it further indents subsequent code?

I do not think that mathlib has specific coding-style rules for `{ ... }`, so that both ways are ok.

> (ii) Is there an automatic way to convert a tactic style proof to a term style proof (with tactics listed in the "(by ...)" after the fact (perhaps in simple cases)? Eg, it's hard for me to image discovering your definition of "invFun" without being in tactic mode.

I am not able either to write the term definition directly! For example, in order to fill the `toFun := fun u => ...`, my first version was probably something like:
```lean
      toFun := fun u ↦ by
        refine' Over.homMk _ sorry -- or apply Over.homMk
        dsimp
        refine' pullback.lift _ _ sorry
        · exact u.left
        · exact x.hom
```
and then, I turned it manually into a term. There is actually a tactic `show_term` which would do the job:
```lean
      toFun := fun u ↦ by
        show_term
        refine' Over.homMk _ sorry
        refine' pullback.lift _ _ sorry
        · exact u.left
        · exact x.hom
``` 
This displays a clickable `Over.homMk (pullback.lift u.left x.hom ⋯) ⋯`.

> (iii) I only have a vague understanding of name spaces. In addition to the renaming you suggest, are you suggesting I move this definition and the previous one to the "Over" file?

The definitions should stay in the same file `Mathlib/CategoryTheory/Limits/Shapes/Pullbacks.lean`, but the full name of the definition should be `CategoryTheory.Over.mapAdjunction` instead of `CategoryTheory.Limits.*` (in the syntax, filenames/location in folders are independent of namespaces, even though most of the definitions named `CategoryTheory....` are in the `Mathlib/CategoryTheory` folder). Entering/exiting namespaces are done using commands `namespace/end`. Here, inserting `end Limits` and `namespace Over` before the definition of `baseChange` should do the work.

**Comment by emilyriehl** [comment]
Thanks for all these helpful explanations.

I've attempted the namespace move but am not sure I got everything exactly right. 

Here is the block ending Limits and opening Over. In particular, in Over, I then opened Limits and redeclared the universe and category variables. 

```
end Limits

namespace Over

open Limits

universe u v
variable {C : Type u} [Category.{v} C]
```

Then below the definition of baseChange I had to change something in "align" that I don't understand
```
#align category_theory.limits.base_change CategoryTheory.Over.baseChange
```
(changing "Limit" to "Over" exactly once)

There is no corresponding "align" after the definition of mapAdjunction. The file the ends by "end Over".

**Comment by adamtopaz** [comment]
In case you find it helpful, here is a slightly golfed version of your adjunction:
```lean
def mapAdjunction [HasPullbacks C] {X Y : C} (f : X ⟶ Y) : Over.map f ⊣ baseChange f :=
  .mkOfHomEquiv <| {
    homEquiv := fun X Y => {
      toFun := fun g => Over.homMk (pullback.lift g.left X.hom <| by simp)
      invFun := fun g => Over.homMk (g.left ≫ pullback.fst) <|
        by simp [← Over.w g, pullback.condition]
      left_inv := by aesop_cat
      right_inv := by
        intro g
        ext
        dsimp
        ext
        · simp
        · simpa using Over.w g |>.symm
    }
  }
```

Also, I think the tag `@[simps! homEquiv_apply homEquiv_symm_apply unit_app counit_app]` should be added to this definition. If you write `simps!?` instead of `simps!` that will show you what lemmas would be automatically added as a result of this.

**Comment by joelriou** [comment]
The very first declaration of universes should be moved before the line `namespace CategoryTheory.Limits` so that we do not need to redeclare these.

The #align mysterious line is a compatibility layer which was very useful when we ported mathlib from Lean 3 to Lean 4: it keeps track of the correspondence between declarations in mathlib3 and mathlib(4): changing `Limits` to `Over` was the right thing to do!

As we are changing the full name of the definition `baseChange`, the best practice would be to add the following lines after the declaration:
```lean
-- deprecated on 2024-05-15
@[deprecated] noncomputable alias Limits.baseChange := Over.baseChange
```
Then, users shall get a warning if they try to use `Limits.baseChange` instead of `Over.baseChange`.

File: Mathlib/CategoryTheory/Limits/Shapes/Pullbacks.lean [hunk 1]
```diff
@@ -2746,6 +2753,27 @@ def baseChange [HasPullbacks C] {X Y : C} (f : X ⟶ Y) : Over Y ⥤ Over X wher
     apply Over.OverMorphism.ext; apply pullback.hom_ext
     · dsimp; simp
     · dsimp; simp
-#align category_theory.limits.base_change CategoryTheory.Limits.baseChange
-
-end CategoryTheory.Limits
+#align category_theory.limits.base_change CategoryTheory.Over.baseChange
+
+/-- The adjunction `Over.map ⊣ baseChange` -/
+
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Limits/Shapes/Pullbacks.lean:2759]: We usually do not put empty lines between the docstring comment and the declaration:
```suggestion
```

**Comment by emilyriehl** [comment]
Hey @adamtopaz. Just added the simps you suggested but have no idea what you mean by "golfed" ;)

**Comment by emilyriehl** [comment]
Happy to add the deprecation warning but I'm not sure exactly where it should go. After the mysterious align line?

**Comment by emilyriehl** [comment]
My local `lake build` succeeds now --- apologies for not checking that before. But I'm getting two errors related to the Pullback file about various things not being in "simp normal form". Advice welcome!

**Comment by joelriou** [comment]
This has to do with the `simps` attribute (which we should keep, but maybe after some tweaks). I will try to investigate tomorrow...

**Comment by adamtopaz** [comment]
It seems that the simp lemmas `Adjunction.homEquiv_unit` and `Adjunction.homEquiv_counit` are firing to simplify the homEquiv's away in two of the automatically generated simps lemmas, causing the simpNF linter to (correctly) complain. Changing the attribute to `@[simps! unit_app counit_app]` should fix this, and since these general lemmas will fire in `simp`, there shouldn't be any harm in removing the simps lemmas for `homEquiv`.

Concerning "golfing", that's just a silly term for "making things shorter". The proofs in my comment above are essentially the same, except a bit shorter: 
```lean
          invFun := fun v => Over.homMk (v.left ≫ pullback.fst) (by
            dsimp
            rw [Category.assoc, pullback.condition, ← Over.w v, Category.assoc]
            dsimp)
```
became
```lean
      invFun := fun g => Over.homMk (g.left ≫ pullback.fst) <|
        by simp [← Over.w g, pullback.condition]
```
and
```lean
          right_inv := fun f => by
            dsimp
            ext
            dsimp
            ext
            . simp
            · dsimp
              rw [pullback.lift_snd, ← Over.w f]
              dsimp } }
```
became
```lean
      right_inv := by
        intro g
        ext
        dsimp
        ext
        · simp
        · simpa using Over.w g |>.symm
    }
  }
```

It's not a huge deal. I mostly mentioned it because in the category theory library we try to do as much work as possible with `aesop_cat`, `ext` and `(d)simp(a)`. This also has some benefits for code maintenance. For example, in the modified proofs above, it's now clear that essentially the only reason why `aesop_cat` couldn't prove everything is because one needs the reversed version of `Over.mk`, and that `pullback.condition` is not a simp lemma, and there's not much we can do about either of these.

**Comment by semorrison** [comment]
By the way, if you want to be able to reproduce the linter error locally, `make lint` should do it. However be warned it is quite slow (will take 15 minutes or more, depending on your machine).

File: Mathlib/CategoryTheory/Limits/Shapes/Pullbacks.lean [hunk 2]
```diff
@@ -2746,6 +2752,29 @@ def baseChange [HasPullbacks C] {X Y : C} (f : X ⟶ Y) : Over Y ⥤ Over X wher
...
-#align category_theory.limits.base_change CategoryTheory.Limits.baseChange
-
-end CategoryTheory.Limits
+#align category_theory.limits.base_change CategoryTheory.Over.baseChange
+
+-- deprecated on 2024-05-15
+@[deprecated] noncomputable alias Limits.baseChange := Over.baseChange
+
+/-- The adjunction `Over.map ⊣ baseChange` -/
+@[simps! homEquiv_apply homEquiv_symm_apply unit_app counit_app]
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Limits/Shapes/Pullbacks.lean:2761]: As suggested by Adam, this should fix the issue:
```suggestion
@[simps! unit_app counit_app]
```

The verification "linting" that is done is a global verification at the scale of all mathlib, which can be very long as Kim said. However, it is possible to make part of these verification on a single file by inserting "#lint".

After this change, this PR should be ready to be merged. Then, you should edit the title of the PR as "feat(CategoryTheory): the adjunction between Over.map and baseChange" (following the conventions https://leanprover-community.github.io/contribute/commit.html)

File: Mathlib/CategoryTheory/Limits/Shapes/Pullbacks.lean [hunk 3]
```diff
@@ -2746,6 +2752,29 @@ def baseChange [HasPullbacks C] {X Y : C} (f : X ⟶ Y) : Over Y ⥤ Over X wher
...
+          right_inv := fun v ↦ by
+            dsimp
+            ext
+            dsimp
+            ext
+            · simp
+            · rw [pullback.lift_snd, ← Over.w v]
+              dsimp }}
+
+end Over
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Limits/Shapes/Pullbacks.lean:2780]: This is not required by the syntax, but it is slightly better to "close" the namespace.
```suggestion
end Over

end CategoryTheory
```

**Comment by emilyriehl** [comment]
Thanks all. I just committed the changes suggested by @joelriou and pushed @adamtopaz's golfed version for efficiency's sake.

**Comment by joelriou** [comment]
Thanks!

bors merge

**Pull request closed without merging**